exit.go

```diff
@@ -374,35 +407,72 @@ func (te *TunaExit) getServiceConn(addr *net.UDPAddr, connID []byte, serviceID b
 
 func (te *TunaExit) listenUDP(port int) error {
 	var err error
-	te.udpConn, err = net.ListenUDP("udp", &net.UDPAddr{Port: port})
+	conn, err := net.ListenUDP("udp", &net.UDPAddr{Port: port})
 	if err != nil {
 		log.Println("Couldn't bind listener:", err)
 		return err
 	}
-	te.readUDP()
+	encConn := udpconn.NewEncryptUDPConn(conn)
+	udpConn, err := te.wrapUDPConn(encConn, nil, nil, nil)
+	if err != nil {
+		log.Println("wrap udp conn err:", err)
+		return err
+	}
+	te.udpConn = udpConn
+	te.readUDP(udpConn)
 	return nil
 }
 
-func (te *TunaExit) readUDP() {
+func (te *TunaExit) readUDP(conn udpconn.Conn) {
 	go func() {
-		clientBuffer := make([]byte, 2048)
+		bufferSize := udpconn.MaxUDPBufferSize
+		if !te.IsServer {
+			bufferSize = te.Service.UDPBufferSize
+		}
+		clientBuffer := make([]byte, bufferSize)
+		err := conn.SetWriteBuffer(bufferSize)
+		if err != nil {
+			return
+		}
+		err = conn.SetReadBuffer(bufferSize)
+		if err != nil {
+			return
+		}
+		connMetadata := new(pb.ConnectionMetadata)
+		var k string
 		for {
-			n, addr, err := te.udpConn.ReadFromUDP(clientBuffer)
+			n, addr, err := conn.ReadFromUDP(clientBuffer)
 			if err != nil {
 				log.Println("Couldn't receive data from client:", err)
 				if strings.Contains(err.Error(), "use of closed network connection") {
 					return
 				}
 				continue
 			}
-			serviceConn, err := te.getServiceConn(addr, clientBuffer[0:2], clientBuffer[2], clientBuffer[3])
+			if bytes.Equal(clientBuffer[:udpconn.PrefixLen], []byte{0, 0, 0, 0}) && !te.Reverse {
+				err, connMetadata = readUDPConnMetadata(conn, addr, te.encryptKeys, clientBuffer[udpconn.PrefixLen:n])
+				if err != nil {
+					log.Println("Couldn't read udp metadata from client:", err)
+					return
+				}
+				k = string(append(connMetadata.PublicKey, connMetadata.Nonce...))
+				continue
+			}
+
+			serviceConn, err := te.getServiceConn(addr, clientBuffer[0:2], clientBuffer[2], clientBuffer[3], k)
 			if err != nil {
+				log.Println("get service conn error:", err)
 				continue
 			}
-			_, err = serviceConn.Write(clientBuffer[4:n])
+			n, _, err = serviceConn.WriteMsgUDP(clientBuffer[udpconn.PrefixLen:n], nil, nil)
```

Comment:
What if a data packet arrives before the metadata packet?

Reply:
This packet can not being decrypt and send to the right local conn

> What if a data packet arrives before the metadata packet?